proposals/p2287.md

```diff
@@ -0,0 +1,116 @@
+# Allow unqualified name lookup for class members
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/####)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+    -   [Namespaces](#namespaces)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [No unqualified lookup when defining outside a scope](#no-unqualified-lookup-when-defining-outside-a-scope)
+
+<!-- tocstop -->
+
+## Abstract
+
+Allow unqualified name lookup for class members, whether inside the class scope
+or within an out-of-line function definition.
+
+## Problem
+
+[Member access](/docs/design/expressions/member_access.md) defines certain
+member access behaviors. However, it doesn't cover what happens if an
+unqualified name lookup occurs within a class, particularly for an out-of-line
+member function definition.
+
+## Background
+
+The [member access design](/docs/design/expressions/member_access.md) and
+[information accumulation principle](/docs/project/principles/information_accumulation.md)
+affect this.
+
+This will also work similarly to
+[unqualified lookup within C++](https://en.cppreference.com/w/cpp/language/unqualified_lookup).
+
+## Proposal
+
+Allow unqualified name lookup for class members.
+
+This proposal updates [the class design](/docs/design/classes.md) to address
+this.
+
+### Namespaces
+
+More generally, this should also be true of other scopes used in declarations.
+In particular, namespaces should also follow the same rule. However, since
+[name lookup](/docs/design/name_lookup.md) has not been fleshed out, this won't
+make much of an update to it.
+
+An example for namespaces would be:
+
+```carbon
+namespace Foo;
+var Foo.a: i32 = 0;
+
+class Foo.B {}
+
+// `b` and `a` are valid here because unqualified name lookup occurs within `Foo`.
+fn Foo.C(B b) -> i32 {
+  return a;
+}
+```
+
+## Rationale
+
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Performing unqualified name lookup for class members should be fairly
+        unsurprising to readers, and should allow for more concise code when
+        working within a namespace.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   This behavior will be similar to how C++ works.
+
+## Alternatives considered
```

Comment:
I would like us to consider the alternative of treating `me` as an ordinary parameter within method bodies, meaning that (modulo aliasing) it can be accessed only by explicitly naming it. In C++, I have often found it to be a readability obstacle that you can't easily find the places in a method body that access `*this`. Treating `me` as an explicitly-declared parameter gives us a clear path to avoiding that problem, so if we're not going to take it, I think we should explain why.

Reply:
I don't think this changes the use of `me` at all? My impression was that this only applies to unqualified lookup that *doesn't* use the instance at all (or `*this` from C++), and only uses the type's *scope*.

Reply:
It would seem like a good idea for the proposal to say something about this to avoid confusion. My interpretation is that given:

```
class X {
  var v: i32;
  fn F[me: Self]() {
    v = 1;
    X.v = 1;
    me.v = 1;
    me.(v) = 1;
    me.(X.v) = 1;
  }
}
```

- For `v =1;`, unqualified lookup finds `X.v`, so this is equivalent to `X.v = 1;`
- For `X.v = 1;`, we are referring to an instance member without an instance, [which is an error](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/member_access.md#non-instance-members).
- For `me.v = 1;`, we perform instance binding and assign to the member `v` of `me`.
- For `me.(v) = 1;`, unqualified lookup finds `X.v`, so this is equivalent to `me.(X.v) = 1;`
- For `me.(X.v) = 1;`, we perform instance binding and assign to the `X` member `v` of `me`.

Reply:
Added the missing part of zygoloid's example (see 981-985 in classes.md).

Note, my intent was just to fix out-of-date information and address what I thought would be a non-contentious change. I don't see `me` as either:

1. I'm not familiar with the rationale behind disallowing `me.` references, nor could I find it considered anywhere.

2. Requiring `me.` is consistent with C++ and seems more ergonomic to write (I don't think it's been a source of confusion in C++ code). #1974 also offers a way to make it work: `me` could be an implicit scoped parameter.

I'm not readily familiar with the argument for _not_ supporting lookups with `me`. There may be refactoring reasons for disallowing it, but I would think those would be shared with unqualified name lookup in general.

Because the issue hasn't been addressed previously and shares problems with unqualified name lookup on classes, I think it's fair to ask here. That said, I also don't feel comfortable resolving this, so I've added notes for an open question.

Is the open question a reasonable answer for now?

Reply:
Ah, sorry, I overlooked the distinction between name lookup and instance binding, so I thought this was proposing to e.g. make `v = 1;` be valid and equivalent to `me.v = 1` in @zygoloid's example. I've added a suggestion to the abstract that would have cleared up my confusion. With that change or something like it, I don't think we even need the open question (but it doesn't hurt to have it).

Reply:
I would stress either way, the open question seems fine.